
<!-- README.md is generated from README.Rmd. Please edit that file -->

# mycoBinR

<!-- badges: start -->

<!-- badges: end -->

The goal of mycoBinR is to is to organize and manage file paths for
outputs generated by a Snakemake pipeline, simplifying downstream data
processing for fungal genome projects.

## Installation

You can install the development version of mycoBinR from
[GitHub](https://github.com/) with:

``` r
# install.packages("devtools")
devtools::install_github("taui11/mycoBinR")
```

## Example

The `create_filepaths_df()` function builds a data frame listing the
file names and their paths for a given project directory, project
number, and project title.

``` r
library(mycoBinR)

filepaths <- create_filepaths_df(
  base_path = "/project/data_output",
  project_nr = "pr_01_",
  TITLE = "001"
)
print(filepaths)
#>              name
#> 1        assembly
#> 2        coverage
#> 3           busco
#> 4        taxonomy
#> 5   assembly_info
#> 6    fasta_folder
#> 7      bed_folder
#> 8        bam_file
#> 9   telomere_left
#> 10 telomere_right
#>                                                                                                               path
#> 1                                                   /project/data_output/flye_output/pr_01_001_flye/assembly.fasta
#> 2                                          /project/data_output/coverm_output/pr_01_001_mean_cov/mean_coverage.tsv
#> 3                            /project/data_output/busco_output/pr_01_001_busco/run_ascomycota_odb12/full_table.tsv
#> 4                                  /project/data_output/diamond_output/pr_01_001_taxonomy_prots/001_taxonomy_prots
#> 5                                                /project/data_output/flye_output/pr_01_001_flye/assembly_info.txt
#> 6                                                     /project/data_output/binning_output/pr_01_001_binned/contigs
#> 7                                                        /project/data_output/binning_output/pr_01_001_binned/beds
#> 8  /project/data_output/coverm_output/pr_01_001_mean_cov/bam_cache/assembly.fasta.pr_01_001.hifireads.fastq.gz.bam
#> 9                                     /project/data_output/telfinder_output/pr_01_001_telomeres/NCR_left_score.txt
#> 10                                   /project/data_output/telfinder_output/pr_01_001_telomeres/NCR_right_score.txt
```
